src/operator/contrib/tvmop/ufunc.cc

```diff
@@ -37,29 +38,88 @@ namespace op {
 
 static constexpr char func_vadd_cpu[] = "vadd";
 static constexpr char func_vadd_gpu[] = "cuda_vadd";
+static constexpr char func_bakcward_vadd_cpu[] = "backward_vadd";
+static constexpr char func_bakcward_vadd_gpu[] = "cuda_backward_vadd";
 
 template<const char* func>
-void TVMBroadcastCompute(const nnvm::NodeAttrs& attrs,
-                         const mxnet::OpContext& ctx,
-                         const std::vector<TBlob>& inputs,
-                         const std::vector<OpReqType>& req,
-                         const std::vector<TBlob>& outputs) {
+void TVMBinaryCompute(const nnvm::NodeAttrs& attrs,
+                      const mxnet::OpContext& ctx,
+                      const std::vector<TBlob>& inputs,
+                      const std::vector<OpReqType>& req,
+                      const std::vector<TBlob>& outputs) {
   CHECK_EQ(inputs.size(), 2U);
   CHECK_EQ(outputs.size(), 1U);
   tvm::runtime::TVMOpModule::Get()->Call(func, ctx, {inputs[0], inputs[1], outputs[0]});
 }
 
+template<const char* func>
+void TVMBinaryBackwardComputeUseNone(const nnvm::NodeAttrs& attrs,
+                                     const mxnet::OpContext& ctx,
+                                     const std::vector<TBlob>& inputs,
+                                     const std::vector<OpReqType>& req,
+                                     const std::vector<TBlob>& outputs) {
+  CHECK_EQ(inputs.size(), 1U);
+  CHECK_EQ(outputs.size(), 2U);
+  int ndim = inputs[0].shape_.ndim();
+  for (int k = 0; k < 2; ++k) {
+    // dispatch by backward
+    std::vector<int> ov, iv;
+    const TBlob& ograd = inputs[0], igrad = outputs[k];
+    bool flag = ograd.size(0) != igrad.size(0);
+    for (int i = 0; i < ndim; ++i) {
```

Comment:
If my understanding is correct, there seems to be an assumption that `ograd.ndim = igrad.ndim`, which is not necessarily true. I think you need to prepend axes before `igrad` if `igrad.ndim < ograd.ndim` and then use the logic here.

Reply:
Yes, `igrad.ndim = ograd.ndim` is assumed.

@yzhliu suggests padding the input to 5-dim, which is the largest possible dim supported by this op. The padding will 1) reduce the number of kernels 2) handle the `igrad.ndim < ograd.ndim` issue. But there may be loss in performance.

I think prepending axes before `igrad` to make it `ograd.dim` requires more kernels, but the performance is better. It is a tradeoff.

Reply:
Please correct me if my understanding is wrong, but don't you still need kernels generated for `ndims < 5` since you will collapse consecutive dimensions where reduction is performed? For example, given a 5d shape `(2, 3, 4, 5, 6)`, and perform reduction on `axis=(1, 2)`, the `tblob` will be first reshaped into `(2, 12, 30)`, and then reduce on `axis=1`.  In this case, do you need a kernel generated for 3D shapes?

Reply:
I think we can pad the shape after dimension collapse. In this case, the `tblob` will be reshaped into `(2, 12, 30, 1, 1)` and then reduce on `axis=[1, 3]`.

Reply:
I see. I am in favor of the approach with less kernels generated. We can revisit the performance concern if that turns out to be an issue.

Reply:
I pushed a new version, where the inputs and outputs are padded to 5 dim.